datafusion/physical-expr/src/scalar_function.rs

```diff
@@ -146,11 +146,22 @@ impl PhysicalExpr for ScalarFunctionExpr {
         // evaluate the function
         match self.fun {
             ScalarFunctionDefinition::UDF(ref fun) => {
-                if self.args.is_empty() {
-                    fun.invoke_no_args(batch.num_rows())
-                } else {
-                    fun.invoke(&inputs)
+                let output = match self.args.is_empty() {
+                    true => fun.invoke_no_args(batch.num_rows()),
+                    false => fun.invoke(&inputs),
+                }?;
+
+                if fun.validate_number_of_rows() {
+                    let output_size = match &output {
+                        ColumnarValue::Array(array) => array.len(),
+                        ColumnarValue::Scalar(_) => 1,
```

Comment:
I don't think this logic is correct -- specifically `ColumnarValue::Scalar` represents (logically) a single value for all the rows.

Thus I think if the function returns `ColumnarValue::Scalar` that implicitly can be any number of rows. 

Therefore I think the check could be something like 
```rust
if let ColumnarValue::Array(array) = &output {
  if output_size != array.len() {
    return internal_err...
  }
}
```

I'll try and make a PR to improve the documentation on ColumnarValue to make this clearer: https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.ColumnarValue.html

Reply:
ah yes, so we don't need to check the length if the output is a scalar value, we actually had a discussion here: https://github.com/apache/datafusion/pull/10148#discussion_r1573629188

Reply:
Here is a PR to improve the docs https://github.com/apache/datafusion/pull/10265

Reply:
oh then perhaps the current implementation of arrow_typeof can be wrong, if we have a map like this:
```
{id:1,name:"bob"}
{id:"string_id",name:"alice"}
```

Reply:
I see. I agree that we can check with Array only.

Reply:
why is map key allowed to have different types

Reply:
Aha i 'v read duckdb's map definition again
```
A dictionary of multiple named values, each key having the same type and each value having the same type. Keys and values can be any type and can be different types from one another
```
I am new to this kind of project, but i think we don't have and should not have that use case 🤔. Else it will affect alot of execution logic

Reply:
In Duckdb, map keys can be **different value** for each row, but the types should not be different.
In OLAP which has the columnar format (we use arrow), it does not make sense to have different types on different rows.

Reply:
> MAPs must have a single type for all keys, and a single type for all values

Reply:
:+1: